File: server/chemistry.py
"""
WandPool Chemistry & Dosing Engine
Baseado nos modelos de cálculo de LSI (Langelier Saturation Index) padrão NSPF / Orenda / Skimmer.
"""

import math
from typing import Dict, Any, List, Optional

def get_temperature_factor(temp_c: float) -> float:
    """Fator de Temperatura (TF) padrão NSPF/LSI."""
    temp_f = (temp_c * 9.0 / 5.0) + 32.0
    # Valores de referência: 10°C (50°F) = 0.35, 20°C (68°F) = 0.55, 25°C (77°F) = 0.68, 30°C (86°F) = 0.76, 35°C (95°F) = 0.84
    # Ajuste por fórmula logarítmica padrão
    tf = (math.log10(max(34.0, temp_f)) * 1.5) - 2.15
    return max(0.0, tf)

def calculate_lsi(
    ph: float,
    temperature_c: float = 25.0,
    calcium_hardness_ppm: float = 250.0,
    total_alkalinity_ppm: float = 100.0,
    cyanuric_acid_ppm: float = 30.0,
    total_dissolved_solids_ppm: float = 1000.0
) -> Dict[str, Any]:
    """
    Calcula o Índice de Saturação de Langelier (LSI) ajustado para Ácido Cianúrico e TDS.
    LSI = pH + TF + CF + AF - Constant
    """
    temp_f = (temperature_c * 9.0 / 5.0) + 32.0
    tf = get_temperature_factor(temperature_c)
    
    # Calcium Factor: CF = log10(CH) - 0.4
    ch = max(10.0, calcium_hardness_ppm)
    cf = math.log10(ch) - 0.4
    
    # Alkalinity Factor: AF = log10(TA - (CYA * 0.33))
    cya_correction = cyanuric_acid_ppm * 0.33
    carbonate_alkalinity = max(10.0, total_alkalinity_ppm - cya_correction)
    af = math.log10(carbonate_alkalinity)
    
    # TDS Constant
    tds_constant = 12.1
    if total_dissolved_solids_ppm > 1500:
        tds_constant = 12.2
    if total_dissolved_solids_ppm > 3000:
        tds_constant = 12.3
        
    lsi = round(ph + tf + cf + af - tds_constant, 2)
    
    # Status e Diagnóstico em Português
    if lsi < -0.30:
        status = "Corrosiva / Agressiva"
        status_code = "corrosive"
        description = (
            "A água está insaturada e corrosiva. Risco de corrosão em motores, tubulações metálicas "
            "e desgaste acelerado do rejunte e revestimento da piscina."
        )
        recommendation = "Eleve o pH com Barrilha ou a Alcalinidade com Bicarbonato."
        badge_color = "red"
    elif lsi > 0.30:
        status = "Incrustante / Saturada"
        status_code = "scaling"
        description = (
            "A água está supersaturada de carbonato de cálcio. Risco de incrustações "
            "nas paredes, aquecedores, placas da célula de sal e água turva."
        )
        recommendation = "Reduza o pH ou a Alcalinidade com redutor de pH/Ácido Muriático."
        badge_color = "amber"
    else:
        status = "Equilibrada / Perfeita"
        status_code = "balanced"
        description = (
            "A água está em perfeito equilíbrio físico-químico. Protege os banhistas e equipamentos."
        )
        recommendation = "Mantenha os níveis atuais de filtração e dosagem de rotina."
        badge_color = "emerald"

    return {
        "lsi": lsi,
        "status": status,
        "status_code": status_code,
        "description": description,
        "recommendation": recommendation,
        "badge_color": badge_color,
        "factors": {
            "temperature_f": round(temp_f, 1),
            "temperature_factor": round(tf, 2),
            "calcium_factor": round(cf, 2),
            "alkalinity_factor": round(af, 2),
            "carbonate_alkalinity": round(carbonate_alkalinity, 1),
            "tds_constant": tds_constant
        }
    }

def calculate_pool_volume(
    shape: str,
    length_m: float,
    width_m: float = 0.0,
    diameter_m: float = 0.0,
    shallow_depth_m: float = 1.0,
    deep_depth_m: float = 1.6
) -> Dict[str, Any]:
    """Calcula o volume da piscina em Litros e Galões."""
    avg_depth = (shallow_depth_m + deep_depth_m) / 2.0
    shape_clean = shape.lower().strip()
    
    if shape_clean in ["retangular", "retangulo", "rectangular"]:
        volume_m3 = length_m * width_m * avg_depth
    elif shape_clean in ["redonda", "circular", "round"]:
        radius = diameter_m / 2.0 if diameter_m > 0 else length_m / 2.0
        volume_m3 = math.pi * (radius ** 2) * avg_depth
    elif shape_clean in ["oval"]:
        volume_m3 = math.pi * (length_m / 2.0) * (width_m / 2.0) * avg_depth
    elif shape_clean in ["irregular", "livre", "freeform"]:
        volume_m3 = length_m * width_m * avg_depth * 0.85
    else:
        volume_m3 = length_m * width_m * avg_depth

    liters = round(volume_m3 * 1000.0, 0)
    gallons = round(liters * 0.264172, 0)

    return {
        "shape": shape,
        "avg_depth_m": round(avg_depth, 2),
        "volume_m3": round(volume_m3, 2),
        "liters": int(liters),
        "gallons": int(gallons)
    }

def calculate_chemical_dosages(
    volume_liters: float,
    current_ph: float,
    target_ph: float = 7.4,
    current_fc: float = 1.0,
    target_fc: float = 3.0,
    current_ta: float = 80.0,
    target_ta: float = 100.0,
    current_ch: float = 200.0,
    target_ch: float = 250.0,
    current_cya: float = 20.0,
    target_cya: float = 40.0,
    current_salt: float = 0.0,
    target_salt: float = 0.0,
    pool_type: str = "tradicional"
) -> Dict[str, Any]:
    """Calcula dosagens de produtos químicos recomendados com formatos métrico e US (lbs, fl oz, gal)."""
    volume_gallons = round(volume_liters * 0.264172, 0)
    scale_10k_gal = volume_gallons / 10000.0
    recommendations: List[Dict[str, Any]] = []

    # 1. Ajuste de pH
    delta_ph = round(target_ph - current_ph, 2)
    if delta_ph > 0.1:
        soda_ash_oz = round((delta_ph / 0.1) * 3.0 * scale_10k_gal, 1)
        soda_ash_lbs = round(soda_ash_oz / 16.0, 1)
        formatted_us = f"{soda_ash_lbs} lbs ({soda_ash_oz} oz)" if soda_ash_lbs >= 1 else f"{soda_ash_oz} oz"
        recommendations.append({
            "parameter": "pH (Elevação)",
            "chemical": "Soda Ash / Barrilha Leve (Carbonato de Sódio)",
            "amount": soda_ash_lbs if soda_ash_lbs >= 1 else soda_ash_oz,
            "unit": "lbs" if soda_ash_lbs >= 1 else "oz",
            "amount_formatted": formatted_us,
            "instructions": "Dissolva previamente em um balde com água da piscina e espalhe com a circulação ligada.",
            "priority": "Alta"
        })
    elif delta_ph < -0.1:
        acid_fl_oz = round((abs(delta_ph) / 0.1) * 5.0 * scale_10k_gal, 1)
        acid_qts = round(acid_fl_oz / 32.0, 2)
        formatted_us = f"{acid_qts} qt ({acid_fl_oz} fl oz)" if acid_qts >= 1 else f"{acid_fl_oz} fl oz"
        recommendations.append({
            "parameter": "pH (Redução)",
            "chemical": "Muriatic Acid 31.45% (Ácido Muriático)",
            "amount": acid_fl_oz,
            "unit": "fl oz",
            "amount_formatted": formatted_us,
            "instructions": "ATENÇÃO: Sempre adicione o ácido à água, NUNCA água ao ácido.",
            "priority": "Alta"
        })

    # 2. Ajuste de Alcalinidade Total
    delta_ta = round(target_ta - current_ta, 1)
    if delta_ta >= 10:
        bicarb_lbs = round((delta_ta / 10.0) * 1.5 * scale_10k_gal, 1)
        recommendations.append({
            "parameter": "Alcalinidade Total",
            "chemical": "Sodium Bicarbonate (Bicarbonato de Sódio)",
            "amount": bicarb_lbs,
            "unit": "lbs",
            "amount_formatted": f"{bicarb_lbs} lbs",
            "instructions": "Adicione diluído ao redor da borda da piscina com o filtro recirculando.",
            "priority": "Média"
        })

    # 3. Cloração / Desinfecção
    delta_fc = round(target_fc - current_fc, 1)
    if delta_fc > 0.2:
        liquid_cl_fl_oz = round(delta_fc * 10.7 * scale_10k_gal, 1)
        liquid_gal = round(liquid_cl_fl_oz / 128.0, 2)
        dichlor_oz = round(delta_fc * 2.4 * scale_10k_gal, 1)
        dichlor_lbs = round(dichlor_oz / 16.0, 1)
        
        recommendations.append({
            "parameter": "Cloro Livre (Desinfecção)",
            "chemical": "Liquid Chlorine 12.5% (Hipoclorito de Sódio)",
            "amount": liquid_cl_fl_oz,
            "unit": "fl oz",
            "amount_formatted": f"{liquid_cl_fl_oz} fl oz ({liquid_gal} gal)",
            "alternative": f"Ou {dichlor_lbs if dichlor_lbs >= 1 else dichlor_oz} {'lbs' if dichlor_lbs >= 1 else 'oz'} de Dicloro Granulado 56%",
            "instructions": "Aplique no final da tarde ou à noite para evitar que o sol degrade o cloro.",
            "priority": "Crítica"
        })

    # 4. Dureza Cálcica
    delta_ch = round(target_ch - current_ch, 1)
    if delta_ch >= 20:
        calcium_lbs = round((delta_ch / 10.0) * 1.25 * scale_10k_gal, 1)
        recommendations.append({
            "parameter": "Dureza Cálcica",
            "chemical": "Calcium Chloride (Cloreto de Cálcio)",
            "amount": calcium_lbs,
            "unit": "lbs",
            "amount_formatted": f"{calcium_lbs} lbs",
            "instructions": "Dissolva bem em balde com água antes de aplicar (o produto aquece durante a dissolução).",
            "priority": "Baixa"
        })

    # 5. Ácido Cianúrico
    delta_cya = round(target_cya - current_cya, 1)
    if delta_cya >= 10:
        cya_lbs = round((delta_cya / 10.0) * 0.81 * scale_10k_gal, 1)
        recommendations.append({
            "parameter": "Ácido Cianúrico (Estabilizador)",
            "chemical": "Pool Stabilizer / Conditioner (Ácido Cianúrico)",
            "amount": cya_lbs,
            "unit": "lbs",
            "amount_formatted": f"{cya_lbs} lbs",
            "instructions": "Coloque em um sock no skimmer para dissolução lenta.",
            "priority": "Média"
        })

    # 6. Sal
    if target_salt > 0 and current_salt < target_salt:
        delta_salt = target_salt - current_salt
        salt_lbs = round((delta_salt / 100.0) * 8.33 * scale_10k_gal, 0)
        bags_40lbs = math.ceil(salt_lbs / 40.0)
        recommendations.append({
            "parameter": "Sal para Gerador de Cloro (SWG)",
            "chemical": "Pool Salt (Sal Especial para Piscina 99.8%)",
            "amount": int(salt_lbs),
            "unit": "lbs",
            "amount_formatted": f"{int(salt_lbs)} lbs (~{bags_40lbs} sacos de 40 lbs)",
            "instructions": "Espalhe o sal e varra até dissolver completamente antes de ligar o gerador.",
            "priority": "Alta"
        })

    return {
        "volume_liters": volume_liters,
        "volume_gallons": int(volume_gallons),
        "recommendations": recommendations,
        "total_actions": len(recommendations)
    }

File: client/src/lib/chemistry.ts
import type { LSISolution, DosageResult, ChemicalRecommendation } from '../types/pool';

export function calculateLSIClient(
  ph: number,
  tempInput: number = 78, // default 78°F ou 25°C
  ch: number = 250,
  ta: number = 100,
  cya: number = 30,
  tds: number = 1000
): LSISolution {
  // Se tempInput for menor que 45, assume °C e converte para °F
  const tempF = tempInput < 45 ? (tempInput * 9) / 5 + 32 : tempInput;
  const tf = Math.max(0, Math.log10(Math.max(34, tempF)) * 1.5 - 2.15);
  const cf = Math.log10(Math.max(10, ch)) - 0.4;
  const carbAlk = Math.max(10, ta - cya * 0.33);
  const af = Math.log10(carbAlk);

  let tdsConstant = 12.1;
  if (tds > 1500) tdsConstant = 12.2;
  if (tds > 3000) tdsConstant = 12.3;

  const rawLsi = ph + tf + cf + af - tdsConstant;
  const lsi = Number(rawLsi.toFixed(2));

  let status = 'Equilibrada / Perfeita';
  let statusCode: 'corrosive' | 'balanced' | 'scaling' = 'balanced';
  let badgeColor: 'red' | 'emerald' | 'amber' = 'emerald';
  let description = 'A água está em perfeito equilíbrio físico-químico. Protege banhistas, rejuntes e equipamentos.';
  let recommendation = 'Mantenha os níveis atuais de filtração e dosagem rotineira.';

  if (lsi < -0.30) {
    status = 'Corrosiva / Agressiva';
    statusCode = 'corrosive';
    badgeColor = 'red';
    description = 'A água está ácida/insaturada. Risco de corrosão em motores, aquecedores e desgaste de rejuntes e alvenaria.';
    recommendation = 'Eleve o pH com Barrilha (Soda Ash) ou a Alcalinidade com Bicarbonato de Sódio.';
  } else if (lsi > 0.30) {
    status = 'Incrustante / Saturada';
    statusCode = 'scaling';
    badgeColor = 'amber';
    description = 'A água está supersaturada de cálcio. Risco de incrustações ásperas nas paredes, calcificação em células de sal e turbidez.';
    recommendation = 'Reduza o pH ou a Alcalinidade com Ácido Muriático (Muriatic Acid 31.45%).';
  }

  return {
    lsi,
    status,
    status_code: statusCode,
    description,
    recommendation,
    badge_color: badgeColor,
    factors: {
      temperature_f: Number(tempF.toFixed(1)),
      temperature_factor: Number(tf.toFixed(2)),
      calcium_factor: Number(cf.toFixed(2)),
      alkalinity_factor: Number(af.toFixed(2)),
      carbonate_alkalinity: Number(carbAlk.toFixed(1)),
      tds_constant: tdsConstant
    }
  };
}

export function calculateDosagesClient(
  volumeInput: number, // em Litros ou Galões
  currentPh: number,
  targetPh: number = 7.4,
  currentFc: number = 1.0,
  targetFc: number = 3.0,
  currentTa: number = 80,
  targetTa: number = 100,
  currentCh: number = 200,
  targetCh: number = 250,
  currentCya: number = 20,
  targetCya: number = 40,
  currentSalt: number = 0,
  targetSalt: number = 0,
  isGallons: boolean = false
): DosageResult {
  // Volume normalizado em galões (base padrão dos EUA) e litros
  const volumeGallons = isGallons ? volumeInput : Math.round(volumeInput * 0.264172);
  const volumeLiters = isGallons ? Math.round(volumeInput * 3.78541) : volumeInput;
  const scale10kGal = volumeGallons / 10000;

  const recommendations: ChemicalRecommendation[] = [];

  // 1. pH
  const deltaPh = Number((targetPh - currentPh).toFixed(2));
  if (deltaPh > 0.1) {
    // Soda Ash: ~6 oz por 10k gal para subir ~0.2 pH
    const sodaAshOz = Math.round((deltaPh / 0.1) * 3 * scale10kGal * 10) / 10;
    const sodaAshLbs = Math.round((sodaAshOz / 16) * 10) / 10;
    const formatted = sodaAshLbs >= 1 ? `${sodaAshLbs} lbs (${sodaAshOz} oz)` : `${sodaAshOz} oz`;

    recommendations.push({
      parameter: 'pH (Elevação)',
      chemical: 'Soda Ash / Barrilha Leve (Carbonato de Sódio)',
      amount: sodaAshLbs >= 1 ? sodaAshLbs : sodaAshOz,
      unit: sodaAshLbs >= 1 ? 'lbs' : 'oz',
      amount_formatted: formatted,
      instructions: 'Dissolva previamente em um balde com água da piscina e espalhe com a bomba ligada.',
      priority: 'Alta'
    });
  } else if (deltaPh < -0.1) {
    // Muriatic Acid 31.45%: ~10 fl oz por 10k gal para baixar ~0.2 pH
    const acidFlOz = Math.round((Math.abs(deltaPh) / 0.1) * 5 * scale10kGal * 10) / 10;
    const acidQts = Math.round((acidFlOz / 32) * 100) / 100;
    const formatted = acidQts >= 1 ? `${acidQts} qt (${acidFlOz} fl oz)` : `${acidFlOz} fl oz`;

    recommendations.push({
      parameter: 'pH (Redução)',
      chemical: 'Muriatic Acid 31.45% (Ácido Muriático)',
      amount: acidFlOz,
      unit: 'fl oz',
      amount_formatted: formatted,
      instructions: 'ATENÇÃO: Sempre adicione o ácido à água, NUNCA água ao ácido. Despeje devagar em frente aos retornos.',
      priority: 'Alta'
    });
  }

  // 2. Alcalinidade Total (TA)
  const deltaTa = Number((targetTa - currentTa).toFixed(1));
  if (deltaTa >= 10) {
    // Bicarbonate: ~1.5 lbs por 10k gal para subir 10 ppm TA
    const bicarbLbs = Math.round((deltaTa / 10) * 1.5 * scale10kGal * 10) / 10;
    recommendations.push({
      parameter: 'Alcalinidade Total (TA)',
      chemical: 'Sodium Bicarbonate (Bicarbonato de Sódio)',
      amount: bicarbLbs,
      unit: 'lbs',
      amount_formatted: `${bicarbLbs} lbs`,
      instructions: 'Adicione espalhando ao redor do perímetro mais profundo da piscina com o filtro ligado.',
      priority: 'Média'
    });
  }

  // 3. Cloro Livre (Free Chlorine)
  const deltaFc = Number((targetFc - currentFc).toFixed(1));
  if (deltaFc > 0.2) {
    // Liquid Chlorine 12.5%: ~10.7 fl oz por 10k gal para subir 1 ppm FC
    const liquidClFlOz = Math.round(deltaFc * 10.7 * scale10kGal * 10) / 10;
    const liquidGal = (liquidClFlOz / 128).toFixed(2);
    // Dichlor 56%: ~2.4 oz por 10k gal para subir 1 ppm FC
    const dichlorOz = Math.round(deltaFc * 2.4 * scale10kGal * 10) / 10;
    const dichlorLbs = Math.round((dichlorOz / 16) * 10) / 10;

    recommendations.push({
      parameter: 'Cloro Livre (Sanitização)',
      chemical: 'Liquid Chlorine 12.5% (Hipoclorito de Sódio)',
      amount: liquidClFlOz,
      unit: 'fl oz',
      amount_formatted: `${liquidClFlOz} fl oz (${liquidGal} gal)`,
      alternative: `Ou ${dichlorLbs >= 1 ? `${dichlorLbs} lbs` : `${dichlorOz} oz`} de Dicloro Granulado 56%`,
      instructions: 'Aplique ao entardecer ou à noite para evitar perda imediata de cloro pelos raios UV solares.',
      priority: 'Crítica'
    });
  }

  // 4. Dureza Cálcica (Calcium Hardness)
  const deltaCh = Number((targetCh - currentCh).toFixed(1));
  if (deltaCh >= 20) {
    // Calcium Chloride 77%: ~1.25 lbs por 10k gal para subir 10 ppm CH
    const calciumLbs = Math.round((deltaCh / 10) * 1.25 * scale10kGal * 10) / 10;
    recommendations.push({
      parameter: 'Dureza Cálcica (CH)',
      chemical: 'Calcium Chloride (Cloreto de Cálcio)',
      amount: calciumLbs,
      unit: 'lbs',
      amount_formatted: `${calciumLbs} lbs`,
      instructions: 'Dissolva bem em um balde plástico com água antes de aplicar (o produto aquece ao reagir com a água).',
      priority: 'Baixa'
    });
  }

  // 5. Ácido Cianúrico (CYA / Stabilizer)
  const deltaCya = Number((targetCya - currentCya).toFixed(1));
  if (deltaCya >= 10) {
    // CYA Stabilizer: ~13 oz (0.81 lbs) por 10k gal para subir 10 ppm CYA
    const cyaLbs = Math.round((deltaCya / 10) * 0.81 * scale10kGal * 10) / 10;
    recommendations.push({
      parameter: 'Cyanuric Acid (CYA Estabilizador)',
      chemical: 'Pool Stabilizer / Conditioner (Ácido Cianúrico)',
      amount: cyaLbs,
      unit: 'lbs',
      amount_formatted: `${cyaLbs} lbs`,
      instructions: 'Coloque em uma meia/sock no cesto do skimmer para dissolução lenta sem turvar a água.',
      priority: 'Média'
    });
  }

  // 6. Sal para Piscinas Salinas (SWG)
  if (targetSalt > 0 && currentSalt < targetSalt) {
    const deltaSalt = targetSalt - currentSalt;
    // Salt: ~30 lbs por 10k gal para subir 360 ppm Salt -> ~8.33 lbs por 10k gal por 100 ppm
    const saltLbs = Math.round((deltaSalt / 100) * 8.33 * scale10kGal);
    const bags40lbs = Math.ceil(saltLbs / 40);

    recommendations.push({
      parameter: 'Sal para Gerador SWG (NaCl)',
      chemical: 'Pool Salt (Sal Puro para Piscina 99.8%)',
      amount: saltLbs,
      unit: 'lbs',
      amount_formatted: `${saltLbs} lbs (~${bags40lbs} sacos de 40 lbs)`,
      instructions: 'Espalhe no fundo raso e varra até dissolver completamente antes de reativar a célula geradora de sal.',
      priority: 'Alta'
    });
  }

  return {
    volume_liters: volumeLiters,
    recommendations,
    total_actions: recommendations.length
  };
}

export function calculatePoolVolumeClient(
  shape: string,
  lengthFt: number = 28.0,
  widthFt: number = 14.0,
  diameterFt: number = 18.0,
  shallowDepthFt: number = 3.5,
  deepDepthFt: number = 6.0
) {
  const avgDepthFt = (shallowDepthFt + deepDepthFt) / 2;
  const s = shape.toLowerCase().trim();
  let volumeCuFt = 0;

  if (s === 'retangular' || s === 'retangulo' || s === 'rectangular') {
    volumeCuFt = lengthFt * widthFt * avgDepthFt;
  } else if (s === 'redonda' || s === 'circular' || s === 'round') {
    const r = diameterFt > 0 ? diameterFt / 2 : lengthFt / 2;
    volumeCuFt = Math.PI * r * r * avgDepthFt;
  } else if (s === 'oval') {
    volumeCuFt = Math.PI * (lengthFt / 2) * (widthFt / 2) * avgDepthFt;
  } else if (s === 'irregular' || s === 'livre' || s === 'freeform') {
    volumeCuFt = lengthFt * widthFt * avgDepthFt * 0.85;
  } else {
    volumeCuFt = lengthFt * widthFt * avgDepthFt;
  }

  const gallons = Math.round(volumeCuFt * 7.48052);
  const liters = Math.round(gallons * 3.78541);
  const volumeM3 = Number((liters / 1000).toFixed(1));

  return {
    shape,
    avg_depth_ft: Number(avgDepthFt.toFixed(2)),
    avg_depth_m: Number((avgDepthFt * 0.3048).toFixed(2)),
    volume_cu_ft: Number(volumeCuFt.toFixed(1)),
    volume_m3: volumeM3,
    gallons,
    liters
  };
}
